Teleport crashed with nil pointer error when giving '--insecure-no-tls' option without setting 'tunnel_listen_addr'

### Expected behavior

The program should not crash due to a nil pointer reference error.

### Current behavior

Teleport crashed with nil pointer error when giving '--insecure-no-tls' option without setting 'tunnel_listen_addr' flag. I have attached a config file that reproduces the error.

Bug details:

- Teleport version
  12.1.2

### Recreation steps

Set the `/etc/teleport.yaml` file to the following configuration and run the `teleport start --insecure-no-tls` command.

```yaml
version: v3
teleport:
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: sha256:xxxxxxxxxxxxxxxxxxxx
  diag_addr: ""
  auth_server: my-auth-server:3025
  auth_token: my-token
auth_service:
  enabled: "no"
  proxy_listener_mode: multiplex
proxy_service:
  enabled: "yes"
  https_keypairs: []
  https_keypairs_reload_interval: 0s
  acme: {}
  public_addr: mine.teleport.com:443
  web_listen_addr: "0.0.0.0:443"
  mysql_listen_addr: "0.0.0.0:3306"
```

### Debug logs

```
2023-04-08T12:49:43Z DEBU [PROC:1]    Service has started. pid:29859.1 service:proxy.reversetunnel.server service/supervisor.go:275
2023-04-08T12:49:43Z INFO [PROXY:SER] Reverse tunnel service 12.1.2:api/v12.1.2-0-gbfd4ff0 is starting on . pid:29859.1 utils/cli.go:286
2023-04-08T12:49:43Z INFO [PROXY:SER] Starting 12.1.2:api/v12.1.2-0-gbfd4ff0 on  using in-memory cache pid:29859.1 service/service.go:3500
2023-04-08T12:49:43Z DEBU [PROC:1]    Broadcasting event. event:ProxyReverseTunnelReady pid:29859.1 service/supervisor.go:383
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x168d464]

goroutine 301 [running]:
github.com/gravitational/teleport/lib/limiter.(*Listener).Addr(0x5fc8f7?)
        <autogenerated>:1 +0x24
github.com/gravitational/teleport/lib/sshutils.(*Server).Addr(0x3acde5cb?)
        github.com/gravitational/teleport/lib/sshutils/server.go:312 +0x92
github.com/gravitational/teleport/lib/sshutils.(*Server).acceptConnections(0xc000179600)
        github.com/gravitational/teleport/lib/sshutils/server.go:409 +0x5b
github.com/gravitational/teleport/lib/sshutils.(*Server).Serve(0xc000090fd0?, {0x760a740?, 0xc001504768?})
        github.com/gravitational/teleport/lib/sshutils/server.go:319 +0x69
created by github.com/gravitational/teleport/lib/reversetunnel.(*server).Start
        github.com/gravitational/teleport/lib/reversetunnel/srv.go:592 +0xa5
jenkins@teleport-pro-proxy:~$ sudo service teleport start
```

### Cause of the bug

The code wrapping the reverse tunnel listener is not validating that the embedded listener is nil. 

https://github.com/gravitational/teleport/blob/40c113b8dad017613c154efaa16997a186eec7fb/lib/limiter/listener.go#L34-L39

The wrapped structure does not implement all the methods of net.Listener and delegates some of them to the embedded member.

https://github.com/gravitational/teleport/blob/40c113b8dad017613c154efaa16997a186eec7fb/lib/limiter/listener.go#L26-L29

Because the structure is not nil, it passes the defense code, but when you try to call a method implemented by the embedded member, you encounter a nil pointer reference.

https://github.com/gravitational/teleport/blob/e667486c31144de1f3180062235c3c0859106e92/lib/sshutils/server.go#L337-L340

### How to fix it

Consider modifying the `NewListener` function to look like this:

```go
func NewListener(ln net.Listener, limiter *ConnectionsLimiter) *Listener {
        if ln == nil || limiter == nil {
            return nil
        }
	return &Listener{
		Listener: ln,
		limiter:  limiter,
	}
}
```

This will make the defense code work properly. I'll make a pull request if my suggestion is accepted.

--- a/lib/limiter/limiter.go
+++ b/lib/limiter/limiter.go
@@ -180,7 +180,7 @@ func (l *Limiter) StreamServerInterceptor(srv interface{}, serverStream grpc.Ser
 
 // WrapListener returns a [Listener] that wraps the provided listener
 // with one that limits connections
-func (l *Limiter) WrapListener(ln net.Listener) *Listener {
+func (l *Limiter) WrapListener(ln net.Listener) (*Listener, error) {
 	return NewListener(ln, l.ConnectionsLimiter)
 }
 

--- a/lib/limiter/limiter_test.go
+++ b/lib/limiter/limiter_test.go
@@ -407,7 +407,8 @@ func TestListener(t *testing.T) {
 			limiter, err := NewConnectionsLimiter(test.config)
 			require.NoError(t, err)
 
-			ln := NewListener(test.listener, limiter)
+			ln, err := NewListener(test.listener, limiter)
+			require.NoError(t, err)
 
 			// open connections without closing to enforce limits
 			conns := make([]net.Conn, 0, connLimit)

--- a/lib/limiter/listener.go
+++ b/lib/limiter/listener.go
@@ -35,11 +35,15 @@ type Listener struct {
 // NewListener creates a [Listener] that enforces the limits of
 // the provided [ConnectionsLimiter] on the all connections accepted
 // by the provided [net.Listener].
-func NewListener(ln net.Listener, limiter *ConnectionsLimiter) *Listener {
+func NewListener(ln net.Listener, limiter *ConnectionsLimiter) (*Listener, error) {
+	if ln == nil {
+		return nil, trace.BadParameter("listener cannot be nil")
+	}
+
 	return &Listener{
 		Listener: ln,
 		limiter:  limiter,
-	}
+	}, nil
 }
 
 // Accept waits for and returns the next connection to the listener

--- a/lib/service/service.go
+++ b/lib/service/service.go
@@ -2849,7 +2849,12 @@ func (process *TeleportProcess) initSSH() error {
 			}()
 			defer mux.Close()
 
-			go s.Serve(limiter.WrapListener(mux.SSH()))
+			listener, err = limiter.WrapListener(mux.SSH())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			go s.Serve(listener)
 		} else {
 			// Start the SSH server. This kicks off updating labels and starting the
 			// heartbeat.
@@ -4118,6 +4123,11 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 			}
 		}
 
+		rtListener, err := reverseTunnelLimiter.WrapListener(listeners.reverseTunnel)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
 		tsrv, err = reversetunnel.NewServer(
 			reversetunnel.Config{
 				Context:               process.ExitContext(),
@@ -4125,7 +4135,7 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 				ID:                    process.Config.HostUUID,
 				ClusterName:           clusterName,
 				ClientTLS:             clientTLSConfig,
-				Listener:              reverseTunnelLimiter.WrapListener(listeners.reverseTunnel),
+				Listener:              rtListener,
 				GetHostSigners:        sshutils.StaticHostSigners(conn.ServerIdentity.KeySigner),
 				LocalAuthClient:       conn.Client,
 				LocalAccessPoint:      accessPoint,
@@ -4639,14 +4649,24 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 
 		// start ssh server
 		go func() {
-			if err := sshProxy.Serve(proxyLimiter.WrapListener(listeners.ssh)); err != nil && !utils.IsOKNetworkError(err) {
+			listener, err := proxyLimiter.WrapListener(listeners.ssh)
+			if err != nil {
+				logger.ErrorContext(process.ExitContext(), "Failed to set up SSH proxy server", "error", err)
+				return
+			}
+			if err := sshProxy.Serve(listener); err != nil && !utils.IsOKNetworkError(err) {
 				logger.ErrorContext(process.ExitContext(), "SSH proxy server terminated unexpectedly", "error", err)
 			}
 		}()
 
 		// start grpc server
 		go func() {
-			if err := sshGRPCServer.Serve(proxyLimiter.WrapListener(listeners.sshGRPC)); err != nil && !utils.IsOKNetworkError(err) && !errors.Is(err, grpc.ErrServerStopped) {
+			listener, err := proxyLimiter.WrapListener(listeners.sshGRPC)
+			if err != nil {
+				logger.ErrorContext(process.ExitContext(), "Failed to set up SSH proxy server", "error", err)
+				return
+			}
+			if err := sshGRPCServer.Serve(listener); err != nil && !utils.IsOKNetworkError(err) && !errors.Is(err, grpc.ErrServerStopped) {
 				logger.ErrorContext(process.ExitContext(), "SSH gRPC server terminated unexpectedly", "error", err)
 			}
 		}()

--- a/lib/sshutils/server.go
+++ b/lib/sshutils/server.go
@@ -335,7 +335,12 @@ func (s *Server) Start() error {
 			return trace.ConvertSystemError(err)
 		}
 
-		if err := s.SetListener(s.limiter.WrapListener(listener)); err != nil {
+		listener, err = s.limiter.WrapListener(listener)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := s.SetListener(listener); err != nil {
 			return trace.Wrap(err)
 		}
 	}

